python/tskit/metadata.py

```diff
@@ -446,13 +446,23 @@ def make_object_or_null_encode(cls, sub_schema):
             key: StructCodec.make_encode(prop)
             for key, prop in sub_schema["properties"].items()
         }
-        return (
-            lambda obj: b""
-            if obj is None
-            else b"".join(
-                sub_encoder(obj[key]) for key, sub_encoder in sub_encoders.items()
-            )
-        )
```

Comment:
This was a bug, I think - `make_object_encode` was using defaults, but `make_object_or_null_encode` was not, leading to a `KeyError` when trying to encode objects without all the required properties.